nixos/modules/services/security/reaction.nix

```diff
@@ -0,0 +1,194 @@
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to run reaction as root.
+        Defaults to false, where an unprivileged reaction user is created.
+
+        Be sure to give it sufficient permissions.
+        Example config permitting `iptables` and `journalctl` use
+
+        ```nix
+        {
+          # allows reading journal logs of processess
+          users.users.reaction.extraGroups = [ "systemd-journal" ];
+          # allows modifying ip firewall rules
+          systemd.services.reaction.AmbientCapabilities = [ "CAP_NET_ADMIN" ];
+          # optional, if more control over ssh logs is needed
+          services.openssh.settings.LogLevel = lib.mkDefault "VERBOSE";
+        }
+        ```
+      '';
+    };
+  };
+
+  config =
+    let
+      cfg = config.services.reaction;
+
+      generatedSettings = settingsFormat.generate "reaction.yml" cfg.settings;
+      namedGeneratedSettings = lib.optional (cfg.settings != { }) {
+        name = "reaction.yml";
+        path = generatedSettings;
+      };
+
+      # SAFETY: We can discard the dependencies of "file" in the name attribute because we keep them in the path attribute
+      # See https://nix.dev/manual/nix/2.32/language/string-context
+      namedSettingsFiles = builtins.map (file: {
+        name = builtins.unsafeDiscardStringContext (builtins.baseNameOf file);
+        path = file;
+      }) cfg.settingsFiles;
+
+      settingsDir = pkgs.linkFarm "reaction.d" (namedSettingsFiles ++ namedGeneratedSettings);
+    in
+    lib.mkIf cfg.enable {
+      assertions = [
+        {
+          assertion = cfg.settings != { } || (builtins.length cfg.settingsFiles) != 0;
+          message = "You must specify settings and/or settingsFile options";
+        }
+      ];
+
+      users = lib.mkIf (!cfg.runAsRoot) {
+        users.reaction = {
+          isSystemUser = true;
+          group = "reaction";
+        };
+        groups.reaction = { };
+      };
+
+      system.checks =
+        lib.optional (cfg.checkConfig && pkgs.stdenv.hostPlatform == pkgs.stdenv.buildPlatform)
+          (
+            pkgs.runCommand "reaction-config-validation" { } ''
+              ${lib.getExe cfg.package} test-config -c ${settingsDir}
+              touch $out
+            ''
+          );
+
+      # Easier to debug conf when we have direct access to it,
+      # rather than having to look for it in the systemd service file.
+      environment.etc."reaction".source = settingsDir;
+
+      systemd.services.reaction = {
+        enable = true;
+        description = "Scan logs and take action";
+        after = [ "network.target" ];
+        wantedBy = [ "multi-user.target" ];
+        partOf = lib.optionals cfg.stopForFirewall [ "firewall.service" ];
+        path = [ pkgs.iptables ];
+        unitConfig.ConditionCapability = "CAP_NET_ADMIN";
+        serviceConfig = {
+          Type = "simple";
+          User = if (!cfg.runAsRoot) then "reaction" else "root";
```

Comment:
Can't afford to use `DynamicUser` here because lots of usecases of reaction require giving permissions to the service or the user.